src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -301,57 +282,55 @@ internal unsafe SocketError DoOperationConnect(Socket socket, SafeSocketHandle h
 
         internal unsafe SocketError DoOperationConnectEx(Socket socket, SafeSocketHandle handle)
         {
+            Debug.Assert(_asyncCompletionOwnership == 0, $"Expected 0, got {_asyncCompletionOwnership}");
+
             // ConnectEx uses a sockaddr buffer containing the remote address to which to connect.
             // It can also optionally take a single buffer of data to send after the connection is complete.
             // The sockaddr is pinned with a GCHandle to avoid having to use the object array form of UnsafePack.
             PinSocketAddressBuffer();
 
-            NativeOverlapped* overlapped = AllocateNativeOverlapped();
-            try
+            fixed (byte* bufferPtr = &MemoryMarshal.GetReference(_buffer.Span))
```

Comment:
Is pinning intentional here? `bufferPtr` appears to be unused.

Reply:
Yes.  The overhead of `fixed` is essentially free when compared to the costs of pinning with GCHandle.Alloc, so the pattern here is to use fixed to temporarily pin while we initiate the asynchronous operation, and then only if it completes asynchronously and the object will need to remained pinned until the operation completes do we create a GCHandle.